packages/discord-request/src/manager.ts

```diff
@@ -0,0 +1,330 @@
+		resource: string,
+		init: RequestInit,
+		retries: number
+	) => void;
+};
+
+type Cache = {
+	shutdownSignal?: AbortSignal;
+	queueSweepInterval?: number;
+	bucketSweepInterval?: number;
+};
+
+export type ManagerArgs = Partial<Config> & Cache & Callbacks;
+
+export class Manager {
+	#token: string | null = null;
+	shutdownSignal?: AbortSignal | null;
+	config: Config;
+
+	globalDelay: Promise<void> | null = null;
+	globalReset = -1;
+	globalRequestCounter: number;
+
+	buckets: Map<string, Bucket>;
+	queues: Map<string, Queue>;
+
+	#bucketSweeper = -1;
+	#queueSweeper = -1;
+
+	bucketSweepInterval: number;
+	queueSweepInterval: number;
+
+	onBucketSweep?: Callbacks["onBucketSweep"];
+	onQueueSweep?: Callbacks["onQueueSweep"];
+	onRateLimit?: Callbacks["onRateLimit"];
+	onRequest?: Callbacks["onRequest"];
+
+	constructor({
+		// Request Config
+		api,
+		version,
+		cdn,
+		headers,
+		userAgent,
+		retries,
+		timeout,
+		globalRequestsPerSecond,
+		// Manager Options
+		shutdownSignal,
+		onBucketSweep,
+		onQueueSweep,
+		bucketSweepInterval,
+		queueSweepInterval,
+		onRateLimit,
+		onRequest,
+	}: ManagerArgs) {
+		this.config = {
+			api: api ?? "https://discord.com/api",
+			version: version ?? 10,
+			cdn: cdn ?? "https://cdn.discordapp.com",
+			headers: headers ?? {},
+			userAgent: userAgent ?? `Discord Request v0`,
+			retries: retries ?? 3,
+			timeout: timeout ?? 15 * ONE_SECOND,
+			globalRequestsPerSecond: globalRequestsPerSecond ?? 50,
+		};
+
+		this.buckets = new Map();
+		this.queues = new Map();
+
+		this.globalRequestCounter = this.config.globalRequestsPerSecond;
+
+		this.bucketSweepInterval = bucketSweepInterval ?? 0;
+		this.queueSweepInterval = queueSweepInterval ?? 0;
+		this.onBucketSweep = onBucketSweep;
+		this.onQueueSweep = onQueueSweep;
+		this.onRateLimit = onRateLimit;
+		this.onRequest = onRequest;
+
+		this.shutdownSignal = shutdownSignal;
+		this.shutdownSignal?.addEventListener("abort", this.shutdown);
+
+		this.startSweepers();
+	}
+
+	get globalTimeout() {
+		return this.globalReset + OFFSET - Date.now();
+	}
+
+	startSweepers() {
+		this.#startBucketSweep();
+		this.#startQueueSweep();
+	}
+
+	async queue(data: RequestData) {
+		const route = getRouteInformation(data.path);
+
+		const key = getRouteKey(data.method, route);
+		const bucket = this.#getBucket(key);
+		const queue = this.#getBucketQueue(bucket, route.identifier);
+
+		const { resource, init } = this.#resolve(data);
+
+		return queue.add(route, resource, init);
+	}
+
+	setToken(token: string | null) {
+		this.#token = token;
+		return this;
+	}
+
+	async setGlobalDelay(retryAfter: number) {
+		this.globalDelay = sleep(retryAfter, this.shutdownSignal).then(() => {
+			this.globalDelay = null;
+		});
+
+		return this.globalDelay;
+	}
+
+	setShutdownSignal(signal: AbortSignal) {
+		this.shutdownSignal?.removeEventListener("abort", this.shutdown);
+		this.shutdownSignal = signal;
+		this.shutdownSignal.addEventListener("abort", this.shutdown);
```

Comment:
```suggestion
		this.shutdownSignal.addEventListener("abort", this.shutdown.bind(this));
```

this is called enough that you may want a helper `#boundShutdown` or something for resource saving